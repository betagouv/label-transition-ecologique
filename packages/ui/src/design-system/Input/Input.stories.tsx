import {Meta, StoryObj} from '@storybook/react';
import {action} from '@storybook/addon-actions';

import {Input} from './Input';

const meta: Meta<typeof Input> = {
  title: 'Design System/Input',
  component: Input,
};

export default meta;

type Story = StoryObj<typeof Input>;

/** Sans aucune props renseignée. */
export const Default: Story = {
  args: {},
};

/** Avec une aide à la saisie quand le champ est vide */
export const AvecPlaceholder: Story = {
  args: {
    placeholder: 'Saisir une valeur',
    'data-test': 'ok',
  },
};

/** Avec une valeur */
export const AvecValeur: Story = {
  args: {
    value: 'plop',
    onChange: action('onChange'),
  },
};

/** Avec icône */
export const AvecIcone: Story = {
  args: {
    placeholder: 'placeholder',
    icon: {value: 'search-line'},
  },
};

/** Avec bouton icône */
export const AvecIconeBouton: Story = {
  args: {
    placeholder: 'placeholder',
    icon: {
      buttonProps: {
        icon: 'search-line',
        onClick: action('onClick'),
        title: 'Rechercher',
      },
    },
  },
};

/** Avec un texte à la place de l'icône */
export const AvecIconeTexte: Story = {
  args: {
    placeholder: 'placeholder',
    icon: {text: 'TTC'},
  },
};

/** Variante en fonction de l'état */
export const AvecBordsColorés: Story = {
  args: {
    placeholder: 'placeholder',
    state: 'error',
    icon: {
      buttonProps: {
        icon: 'search-line',
      },
    },
  },
};

/** Variante small */
export const VarianteSmall: Story = {
  args: {
    size: 'sm',
    placeholder: 'placeholder',
  },
};

/** Variante small et bouton icône */
export const VarianteSmallEtBouton: Story = {
  args: {
    size: 'sm',
    placeholder: 'placeholder',
    icon: {
      buttonProps: {
        icon: 'search-line',
        onClick: action('onClick'),
        title: 'Rechercher',
      },
    },
  },
};

/** Variante small et texte à la place de l'icône */
export const VarianteSmallIconeTexte: Story = {
  args: {
    size: 'sm',
    placeholder: 'placeholder',
    icon: {text: 'TTC'},
  },
};
