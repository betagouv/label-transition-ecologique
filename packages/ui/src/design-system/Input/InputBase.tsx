import {Ref, forwardRef} from 'react';
import classNames from 'classnames';
import {Icon, IconValue} from '@design-system/Icon';
import {DefaultButtonProps} from '@design-system/Button/types';
import {Button} from '@design-system/Button';
import {FieldState, stateToTextColor} from '@design-system/Field';
import {preset} from '@tailwind-preset';

// variantes de taille
export type InputSize = 'md' | 'sm';

// types de champ accepté
export type InputType = 'text' | 'password' | 'date' | 'search' | 'tel';

// couleur des bordures en fonction du `state`
const stateToBorderColor: Record<FieldState, string> = {
  default: 'border-grey-4',
  disabled: 'border-grey-4',
  info: 'border-info-1',
  error: 'border-error-1',
  success: 'border-success-1',
  warning: 'border-warning-1',
};

export type InputBaseProps = Omit<
  React.ComponentPropsWithoutRef<'input'>,
  'type'
> & {
  /** Type de saisie */
  type?: InputType;
  /** Valeur courante du champ */
  value?: string;
  /** Taille d'affichage */
  displaySize?: InputSize;
  /** Id pour les tests */
  'data-test'?: string;
  /** Contenu optionnel pour la zone d'icône à droite du champ */
  icon?: IconContent;
  /** Variante en fonction d'un état */
  state?: FieldState;
  /** Pour styler le container */
  containerClassname?: string;
};

type IconContent =
  // affiche un composant `Icon`
  | {value?: IconValue}
  // affiche un texte
  | {text?: string}
  // affiche un composant `Button`
  | {buttonProps?: Omit<DefaultButtonProps, 'variant' | 'size' | 'className'>};

/**
 * Composant de base pour les champs de saisie.
 */
export const InputBase = forwardRef(
  (props: InputBaseProps, ref?: Ref<HTMLInputElement>) => {
    const {
      className,
      containerClassname,
      type = 'text',
      displaySize = 'md',
      icon,
      state,
      ...remainingProps
    } = props;

    const borderColor = stateToBorderColor[state] || stateToBorderColor.default;

    return (
      <div
        className={classNames(
          'inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5',
          borderColor,
          containerClassname
        )}
      >
        <input
          type={type}
          ref={ref}
          className={classNames(
            'grow text-grey-8 px-4 outline-none',
            {
              'text-sm py-2': displaySize === 'sm',
              'text-md py-3': displaySize === 'md',
              'border-r border-solid': !!icon,
              [borderColor]: !!icon,
            },
            className
          )}
          {...remainingProps}
        />
        <InputIconContent {...props} />
      </div>
    );
  }
);

/**
 * Affiche le contenu de la zone icône/texte/bouton à droite du champ.
 */
const InputIconContent = ({
  disabled,
  icon,
  state,
  displaySize: size = 'md',
}: InputBaseProps) => {
  if (!icon) {
    return;
  }

  const iconColor = disabled
    ? 'text-grey-6'
    : stateToTextColor[state] || 'text-primary';

  /** icône */
  if ('value' in icon) {
    return (
      <Icon
        className={classNames('self-center min-w-[3rem]', iconColor)}
        icon={icon.value}
        size={size}
      />
    );
  }

  /** "icône" texte  */
  if ('text' in icon) {
    return (
      <span
        className={classNames(
          'self-center text-xs text-center font-bold min-w-[3rem]',
          iconColor
        )}
      >
        {icon.text}
      </span>
    );
  }

  /** icône bouton */
  if ('buttonProps' in icon) {
    return (
      <Button
        variant="white"
        size={size}
        disabled={disabled}
        className={classNames('rounded-none border-none')}
        // change la couleur de l'icône du bouton pour refléter le `state`
        // TODO: enlever les !important de `Button` pour pouvoir styler le bouton par le className
        style={{color: preset.theme.extend.colors[state]?.['1']}}
        {...icon.buttonProps}
      />
    );
  }
};
