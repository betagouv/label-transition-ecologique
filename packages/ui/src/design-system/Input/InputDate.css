/** cache l'icône du sélecteur de date (fonctionne uniquement pour chrome et probablement edge mais pas pour firefox ni safari) */
.input-date input[type='date']::-webkit-calendar-picker-indicator {
  display: none;
}
