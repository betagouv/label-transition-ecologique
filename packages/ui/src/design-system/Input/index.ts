export * from './Input';
export * from './InputDate';
export * from './InputPassword';
