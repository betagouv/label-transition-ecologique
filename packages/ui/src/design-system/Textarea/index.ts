export * from './Textarea';
export * from './AutoResizedTextarea';
