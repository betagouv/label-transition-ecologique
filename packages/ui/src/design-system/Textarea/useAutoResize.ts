import {RefObject, useLayoutEffect, useRef} from 'react';

/**
 * Rend un champ de texte avec redimensionnement automatique en fonction de son contenu
 * Copié / adapté depuis:
 * https://www.kindacode.com/article/react-typescript-create-an-autosize-textarea-from-scratch/
 * https://css-tricks.com/the-cleanest-trick-for-autogrowing-textareas/#comment-1794830
 */

export const useAutoResize = (
  value?: string,
  forwardRef?: RefObject<HTMLTextAreaElement> | null
) => {
  // Référence locale associée à la balise <textarea/>
  const ref = useRef<HTMLTextAreaElement | null>(null);

  // Référence associée à la balise invisible <div/>
  // Cette div permet de conserver la hauteur pour éviter
  // les sauts dans la page à la mise à jour du texte
  const shadowRef = useRef<HTMLDivElement | null>(null);

  // Référence associée à la balise <textarea>
  // (locale ou issue des props)
  const textareaRef = forwardRef ? forwardRef : ref;

  // Permet de set la taille du textarea au changement de valeur
  useLayoutEffect(() => {
    if (textareaRef && textareaRef.current && shadowRef && shadowRef.current) {
      // Initialise la hauteur du textarea à 0px
      textareaRef.current.style.height = '0px';

      // Met à jour la hauteur en fonction du contenu
      textareaRef.current.style.height =
        textareaRef.current.scrollHeight + 'px';

      // Une div invisible qui permet de conserver la hauteur pour éviter
      // les sauts dans la page à la mise à jour du texte
      shadowRef.current.style.height = textareaRef.current.scrollHeight + 'px';
    }
  }, [value]);

  return {textareaRef, shadowRef};
};
