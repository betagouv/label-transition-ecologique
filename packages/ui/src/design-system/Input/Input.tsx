import {Ref, forwardRef} from 'react';
import {InputBase, InputBaseProps} from './InputBase';
import {InputDate, InputDateProps} from './InputDate';
import {InputSearch, InputSearchProps} from './InputSearch';
import {InputPassword, InputPasswordProps} from './InputPassword';
import {InputNumber, InputNumberProps} from './InputNumber';
import {InputPattern, InputPatternProps} from './InputPattern';
import {InputOTP, InputOTPProps} from './InputOTP';
import {InputTel, InputTelProps} from './InputTel';

type InputProps =
  | (Omit<InputBaseProps, 'type'> & {type: 'text'})
  | (InputNumberProps & {type: 'number'})
  | (InputDateProps & {type: 'date'})
  | (InputSearchProps & {type: 'search'})
  | (InputPasswordProps & {type: 'password'})
  | (InputPatternProps & {type: 'pattern'})
  | (InputTelProps & {type: 'tel'})
  | (InputOTPProps & {type: 'otp'});

/**
 * Affiche un champ de saisie, éventuellement combiné à une zone d'icône (ou de
 * texte) à droite du champ.
 * Toutes les props de l'élément HTML `input` sont acceptées à l'exception des changements suivants :
 * - type : restreint à un sous-ensemble des types standards
 */
export const Input = forwardRef(
  ({type = 'text', ...props}: InputProps, ref?: Ref<HTMLInputElement>) => {
    if (type === 'date') {
      return <InputDate {...(props as InputDateProps)} ref={ref} />;
    }

    if (type === 'search') {
      return <InputSearch {...(props as InputSearchProps)} ref={ref} />;
    }

    if (type === 'password') {
      return <InputPassword {...(props as InputPasswordProps)} ref={ref} />;
    }

    if (type === 'number') {
      return <InputNumber {...(props as InputNumberProps)} ref={ref} />;
    }

    if (type === 'pattern') {
      return <InputPattern {...(props as InputPatternProps)} ref={ref} />;
    }

    if (type === 'otp') {
      return <InputOTP {...(props as InputOTPProps)} ref={ref} />;
    }

    if (type === 'tel') {
      return <InputTel {...(props as InputTelProps)} ref={ref} />;
    }

    return <InputBase {...(props as InputBaseProps)} ref={ref} />;
  }
);
