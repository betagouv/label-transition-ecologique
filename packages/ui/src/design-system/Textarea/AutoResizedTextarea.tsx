import {
  ChangeEvent,
  Ref,
  RefObject,
  forwardRef,
  useEffect,
  useState,
} from 'react';
import { Textarea, TextareaProps } from './Textarea';
import { useAutoResize } from './useAutoResize';

export const AutoResizedTextarea = forwardRef(
  (props: TextareaProps, ref?: Ref<HTMLTextAreaElement>) => {
    const { onChange, value: initialValue, ...restProps } = props;

    // Valeur locale, n√©cessaire pour le refresh des valeurs textareaRef et shadowRef
    const [value, setValue] = useState<string | undefined>(initialValue);

    useEffect(() => setValue(initialValue), [initialValue]);

    const { textareaRef, shadowRef } = useAutoResize(
      value?.toString(),
      ref as RefObject<HTMLTextAreaElement>
    );

    const handleChange = (evt: ChangeEvent<HTMLTextAreaElement>) => {
      onChange?.(evt);
      setValue(evt.target.value);
    };

    return (
      <div className="flex items-start w-full">
        <div ref={shadowRef} className="w-px -mr-px" />
        <Textarea
          {...restProps}
          ref={textareaRef}
          value={value}
          onChange={handleChange}
          resize="none"
        />
      </div>
    );
  }
);
