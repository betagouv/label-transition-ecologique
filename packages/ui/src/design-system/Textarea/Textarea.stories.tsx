import {Meta, StoryObj} from '@storybook/react';
import {action} from '@storybook/addon-actions';
import {Textarea} from './Textarea';
import {useRef, useState} from 'react';
import {Input} from '@design-system/Input';

const meta: Meta<typeof Textarea> = {
  component: Textarea,
  argTypes: {
    displaySize: {
      control: {type: 'select'},
    },
    resize: {
      control: {type: 'select'},
    },
  },
  render: args => {
    const [value, setValue] = useState(args.value);
    const ref = useRef(null);
    return (
      <Textarea
        {...args}
        ref={ref}
        value={value}
        onChange={e => {
          console.log(e);

          action('onChange')(e.target);
          action('ref.current.value')(ref.current.value);
          //   setValue(e);
        }}
      />
    );
  },
};

export default meta;

type Story = StoryObj<typeof Textarea>;

/** Sans aucune props renseignée */
export const Default: Story = {};

/** Avec une aide à la saisie quand le champ est vide */
export const WithPlaceholder: Story = {
  args: {
    placeholder: 'Saisir une valeur',
  },
};

/** Avec une valeur */
export const WithValue: Story = {
  args: {
    value: 'Test',
    onChange: action('onChange'),
  },
};

/** Variante en fonction de l'état */
export const WithColorBorders: Story = {
  render: () => (
    <div className="flex flex-col gap-4">
      <Textarea placeholder="Info" state="info" />
      <Textarea placeholder="Error" state="error" />
      <Textarea placeholder="Success" state="success" />
      <Textarea placeholder="Warning" state="warning" />
    </div>
  ),
};

/** Variante small */
export const SmallVariant: Story = {
  args: {
    placeholder: 'Small',
    displaySize: 'sm',
  },
};
