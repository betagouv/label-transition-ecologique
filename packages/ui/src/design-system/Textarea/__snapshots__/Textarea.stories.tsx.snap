// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Textarea Default smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <textarea class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y">
  </textarea>
</div>
`;

exports[`design-system/Textarea SmallVariant smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <textarea placeholder="Small"
            class="grow text-grey-8 px-4 outline-none  text-sm py-2 resize-y"
  >
  </textarea>
</div>
`;

exports[`design-system/Textarea WithColorBorders smoke-test 1`] = `
<div class="flex flex-col gap-4">
  <div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-info-1">
    <textarea placeholder="Info"
              class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y"
    >
    </textarea>
  </div>
  <div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-error-1">
    <textarea placeholder="Error"
              class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y"
    >
    </textarea>
  </div>
  <div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-success-1">
    <textarea placeholder="Success"
              class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y"
    >
    </textarea>
  </div>
  <div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-warning-1">
    <textarea placeholder="Warning"
              class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y"
    >
    </textarea>
  </div>
</div>
`;

exports[`design-system/Textarea WithPlaceholder smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <textarea placeholder="Saisir une valeur"
            class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y"
  >
  </textarea>
</div>
`;

exports[`design-system/Textarea WithValue smoke-test 1`] = `
<div class="inline-flex items-stretch border border-solid rounded-lg bg-grey-1 overflow-hidden focus-within:border-primary-5 border-grey-4">
  <textarea class="grow text-grey-8 px-4 outline-none  text-md py-3 resize-y">
    Test
  </textarea>
</div>
`;
